baserun/exporter.py

```diff
@@ -0,0 +1,177 @@
+
+        for span in spans:
+            if span.name.startswith("baserun.parent") or not span.name.startswith(
+                "baserun"
+            ):
+                continue
+
+            status = Status(
+                message=span.status.description, code=span.status.status_code.value
+            )
+            prompt_messages = [
+                Message(**message_attrs)
+                for message_attrs in self._extract_prefix_dicts(
+                    span.attributes, SpanAttributes.LLM_PROMPTS
+                )
+            ]
+
+            completions = []
+            for message_attrs in self._extract_prefix_dicts(
+                span.attributes, SpanAttributes.LLM_COMPLETIONS
+            ):
+                if "function_name" in message_attrs:
+                    message_attrs["function_call"] = json.dumps(
+                        {
+                            "name": message_attrs.pop("function_name"),
+                            "arguments": message_attrs.pop("function_arguments"),
+                        }
+                    )
+                completions.append(Message(**message_attrs))
+
+            # Trace IDs are huge integers, so they must be encoded as bytes
+            trace_id_int = span.context.trace_id
+            trace_id = trace_id_int.to_bytes(
+                (trace_id_int.bit_length() + 7) // 8, "big"
+            )
+
+            vendor = span.attributes.get(SpanAttributes.LLM_VENDOR)
+
+            run_str = span.attributes.get(SpanAttributes.BASERUN_RUN)
+            if not run_str:
+                logger.warning("Baserun run attribute not set, cannot submit run")
+
+            run = Baserun.deserialize_run(run_str)
+
+            span_message = Span(
+                run_id=run.run_id,
+                trace_id=trace_id,
+                span_id=span.context.span_id,
+                name=span.name,
+                vendor=vendor,
+                status=status,
+                total_tokens=span.attributes.get(
+                    SpanAttributes.LLM_USAGE_TOTAL_TOKENS, 0
+                ),
+                completion_tokens=span.attributes.get(
+                    SpanAttributes.LLM_USAGE_COMPLETION_TOKENS, 0
+                ),
+                prompt_tokens=span.attributes.get(
+                    SpanAttributes.LLM_USAGE_PROMPT_TOKENS, 0
+                ),
+                prompt_messages=prompt_messages,
+                model=span.attributes.get(SpanAttributes.LLM_REQUEST_MODEL, ""),
+                completions=completions,
+                request_type=span.attributes.get(SpanAttributes.LLM_REQUEST_TYPE, ""),
+                max_tokens=span.attributes.get(SpanAttributes.LLM_REQUEST_MAX_TOKENS),
+                stop=span.attributes.get(SpanAttributes.LLM_CHAT_STOP_SEQUENCES),
+            )
+            span_message.start_time.FromNanoseconds(span.start_time)
+            span_message.end_time.FromNanoseconds(span.end_time)
+
+            set_span_attr(
```

Comment:
Now that all of these attributes are optional, can we remove the whole set_span_attr thing and just set these all on the Span when it is constructed?

Reply:
This is the part that prevents default values from being sent on the span message. (`span_message` is the proto message, `span` is the otel span)

Reply:
Er yeah I'm referring to the span_message on line 77. For example we set total_tokens, completion_tokens, etc. In theory we could just set all properties there, right? Since we've updated the proto to handle undefined fields